source/gateway.go

```
@@ -0,0 +1,549 @@
+	log.Debugf("Adding event handlers for %s", src.rtKind)
+	eventHandler := eventHandlerFunc(handler)
+	src.gwInformer.Informer().AddEventHandler(eventHandler)
+	src.rtInformer.Informer().AddEventHandler(eventHandler)
+	src.nsInformer.Informer().AddEventHandler(eventHandler)
+}
+
+func (src *gatewayRouteSource) Endpoints(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	var endpoints []*endpoint.Endpoint
+	routes, err := src.rtInformer.List(src.rtNamespace, src.rtLabels)
+	if err != nil {
+		return nil, err
+	}
+	gateways, err := src.gwInformer.Lister().Gateways(src.gwNamespace).List(src.gwLabels)
+	if err != nil {
+		return nil, err
+	}
+	namespaces, err := src.nsInformer.Lister().List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	kind := strings.ToLower(src.rtKind)
+	resolver := newGatewayRouteResolver(src, gateways, namespaces)
+	for _, rt := range routes {
+		// Filter by annotations.
+		meta := rt.Metadata()
+		annots := meta.Annotations
+		if !src.rtAnnotations.Matches(labels.Set(annots)) {
+			continue
+		}
+
+		// Check controller annotation to see if we are responsible.
+		if v, ok := annots[controllerAnnotationKey]; ok && v != controllerAnnotationValue {
+			log.Debugf("Skipping %s %s/%s because controller value does not match, found: %s, required: %s",
+				src.rtKind, meta.Namespace, meta.Name, v, controllerAnnotationValue)
+			continue
+		}
+
+		// Get Route hostnames and their targets.
+		hostTargets, err := resolver.resolve(rt)
+		if err != nil {
+			return nil, err
+		}
+		if len(hostTargets) == 0 {
+			log.Debugf("No endpoints could be generated from %s %s/%s", src.rtKind, meta.Namespace, meta.Name)
+			continue
+		}
+
+		// Create endpoints from hostnames and targets.
+		resourceKey := fmt.Sprintf("%s/%s/%s", kind, meta.Namespace, meta.Name)
+		providerSpecific, setIdentifier := getProviderSpecificAnnotations(annots)
+		ttl, err := getTTLFromAnnotations(annots)
+		if err != nil {
```

If we ignore the error here, we'll be passing `nil` to `endpointsForHostname`. Would this even be correct? Isn't it better to return the error?

---

A few sources return the error, but the most common behavior seems to be log a warning and continue. When there's an error, the ttl value is 0 which end up being replaced with whatever the default ttl is configured as.

e.g.: https://github.com/kubernetes-sigs/external-dns/blob/v0.11.1/source/ingress.go#L229-L234